plugin/kprom/kprom.go

```diff
@@ -1,260 +1,299 @@
-		}, []string{"node_id"}),
-
-		writeBytes: factory.NewCounterVec(prometheus.CounterOpts{
-			Namespace: namespace,
-			Name:      "write_bytes_total",
-			Help:      "Total number of bytes written, by broker",
-		}, []string{"node_id"}),
-
-		// read
-
-		readErrs: factory.NewCounterVec(prometheus.CounterOpts{
-			Namespace: namespace,
-			Name:      "read_errors_total",
-			Help:      "Total number of read errors, by broker",
-		}, []string{"node_id"}),
-
-		readBytes: factory.NewCounterVec(prometheus.CounterOpts{
-			Namespace: namespace,
-			Name:      "read_bytes_total",
-			Help:      "Total number of bytes read, by broker",
-		}, []string{"node_id"}),
-
-		// produce & consume
-
-		produceBytes: factory.NewCounterVec(prometheus.CounterOpts{
-			Namespace: namespace,
-			Name:      "produce_bytes_total",
-			Help:      "Total number of uncompressed bytes produced, by broker and topic",
-		}, []string{"node_id", "topic"}),
-
-		fetchBytes: factory.NewCounterVec(prometheus.CounterOpts{
-			Namespace: namespace,
-			Name:      "fetch_bytes_total",
-			Help:      "Total number of uncompressed bytes fetched, by broker and topic",
-		}, []string{"node_id", "topic"}),
-	}
+	return &Metrics{cfg: cfg}, nil
 }
 
-func (m *Metrics) OnNewClient(cl *kgo.Client) {
-	factory := promauto.With(m.cfg.reg)
-
-	factory.NewGaugeFunc(prometheus.GaugeOpts{
-		Namespace: m.cfg.namespace,
-		Name:      "buffered_produce_records_total",
-		Help:      "Total number of records buffered within the client ready to be produced.",
-	}, func() float64 { return float64(cl.BufferedProduceRecords()) })
-
-	factory.NewGaugeFunc(prometheus.GaugeOpts{
-		Namespace: m.cfg.namespace,
-		Name:      "buffered_fetch_records_total",
-		Help:      "Total number of records buffered within the client ready to be consumed.",
-	}, func() float64 { return float64(cl.BufferedFetchRecords()) })
-}
+// interface checks to ensure we implement the hooks properly
+var (
+	_ kgo.HookNewClient           = new(Metrics)
+	_ kgo.HookBrokerConnect       = new(Metrics)
+	_ kgo.HookBrokerDisconnect    = new(Metrics)
+	_ kgo.HookProduceBatchWritten = new(Metrics)
+	_ kgo.HookFetchBatchRead      = new(Metrics)
+	_ kgo.HookBrokerE2E           = new(Metrics)
+)
 
-func strnode(node int32) string {
-	if node < 0 {
-		return "seed_" + strconv.Itoa(int(node)-math.MinInt32)
-	}
-	return strconv.Itoa(int(node))
+func (m *Metrics) OnNewClient(client *kgo.Client) {
+	factory := promauto.With(m.cfg.reg)
+	clientID := client.OptValue(kgo.ClientID).(string)
+	namespace := m.cfg.namespace
+	subsystem := m.cfg.subsystem
+	constLabels := prometheus.Labels{"clientid": clientID}
+	buckets := m.cfg.buckets
+
+	// connects and disconnects
+
+	m.connects = factory.NewCounterVec(prometheus.CounterOpts{
+		Namespace:   namespace,
+		Subsystem:   subsystem,
+		ConstLabels: constLabels,
+		Name:        "connects_total",
+		Help:        "Total number of connections opened, by nodeid",
+	}, []string{"nodeid"})
```

Comment:
Do you prefer `nodeid` over `node_id`, is there a convention for naming the labels without using underscores? I think I'd prefer `node_id` and `client_id` for better readability like it was before.

Reply:
agreed, and updated